perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1922,4 +1964,172 @@ private static void prepareServiceRoles(PerunSession sess) {
+	 * @param objects is a list of PerunBeans against which will be authorization done.
+	 * @return true if the principal has particular rights, false otherwise.
+	 * @throws PolicyNotExistsException when the given policyDefinition does not exist in the PerunPoliciesContainer.
+	 */
+	private static boolean authorized(PerunSession sess, String policyDefinition, List<PerunBean> objects) throws PolicyNotExistsException {
+		// We need to load additional information about the principal
+		if (!sess.getPerunPrincipal().isAuthzInitialized()) {
+			refreshAuthz(sess);
+		}
+
+		// If the user has no roles, deny access
+		if (sess.getPerunPrincipal().getRoles() == null) {
+			return false;
+		}
+
+		List<JsonNode> allPolicies = fetchPolicyWithAllIncludedPolicies(policyDefinition);
+
+		List<JsonNode> policyRoles = new ArrayList<>();
+		for (JsonNode policy : allPolicies) policyRoles.addAll(fetchPolicyRoles(policy));
+
+		//Fetch super objects like Vo for group etc.
+		List<PerunBean> allReliableObjects = fetchAllReliableObjects(objects);
+
+		//create a map from objects for easier manipulation
+		Map <String, List<PerunBean>> mapOfBeans = new HashMap<>();
+		for (PerunBean object: allReliableObjects) {
+			if (!mapOfBeans.containsKey(object.getBeanName())) {
+				List<PerunBean> listOfbeans = new ArrayList<>();
+				listOfbeans.add(object);
+				mapOfBeans.put(object.getBeanName(), listOfbeans);
+			} else mapOfBeans.get(object.getBeanName()).add(object);
+		}
+
+		return resolveAuthorization(sess, policyRoles, mapOfBeans);
+	}
+
+	/**
+	 * Decide whether a principal has sufficient rights according the the given roles and objects.
+	 *
+	 * @param sess perunSession which contains the principal.
+	 * @param policyRoles is a list of JsonNode from configuration file with a list of roles and objects they are connected with or another sub lists of them.
+	 * @param mapOfBeans is a map of objects against which will be authorization done.
+	 * @return true if the principal has particular rights, false otherwise.
+	 */
+	private static boolean resolveAuthorization(PerunSession sess, List<JsonNode> policyRoles, Map <String, List<PerunBean>> mapOfBeans) {
+		//Traverse through outer role list which works like logical OR
+		for (JsonNode roleArray: policyRoles) {
+
+			boolean authorized = true;
+			//Traverse through inner role list which works like logical AND
+			Iterator<String> roleArrayKeys = roleArray.fieldNames();
+			while (roleArrayKeys.hasNext()) {
+				String role = roleArrayKeys.next();
+
+				//fetch the role, object is connected with, from config
+				JsonNode roleObject = roleArray.get(role);
+
+				// If policy role is not connected to any object or if there is no corresponding type of object in the perunBeans map
+				if (roleObject.isNull() || !mapOfBeans.containsKey(roleObject.textValue())) {
+					//If principal does not have the role, this inner list's result is false
+					if (!sess.getPerunPrincipal().getRoles().hasRole(role)) {
```

Comment:
You can probably call only once this method. You do not need to do it in every iteration.

Reply:
getRoles() does not retrieve the roles from the database, it is called from the AuthzRoles object so it is ok.